### PR TITLE
Add Mask2Former + OneFormer blog

### PR DESCRIPTION
**Pull request by NielsRogge**
This PR adds a blog post titled "Towards Universal Image Segmentation with Mask2Former and OneFormer", to introduce the new Mask2Former and OneFormer models, which are now available in 🤗 Transformers.

cc @shivalikasingh95

To do:

- [x] thumbnail

### CODE REVIEW

File: mask2former.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
+</a>
+
+**This guide introduces Mask2Former, a state-of-the-art neural network for image segmentation. The model is now available in [`🤗 transformers`](https://huggingface.co/transformers), an open-source library that offers easy-to-use implementations of state-of-the-art models. Along the way, you'll learn about the difference between the various forms of image segmentation.**
+
+## Image segmentation
+
+Image segmentation is the task of identifying different "segments" in an image, like people or cars. More technically, image segmentation is the task of grouping pixels with different semantics. Refer to the HuggingFace [task page](https://huggingface.co/tasks/image-segmentation) for a brief introduction.
+
+Over the years, people have come up with different ways of performing image segmentation. They can largely be split into 3 subtasks, called instance, semantic and panoptic segmentation:
```
Comment (alaradirik) [inline mask2former.md:39]: Could we rephrase this a bit to emphasize how previous works approached the subtasks separately?
```suggestion
Image segmentation can largely be split into 3 subtasks - instance, semantic and panoptic segmentation - with numerous methods and model architectures to perform each subtask.
```

File: mask2former.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+Over the last years, researchers have come up with several architectures that were typically very tailored to either instance, semantic or panoptic segmentation. Instance and panoptic segmentation were typically solved by outputting a set of binary masks + corresponding labels (very similar to object detection, except that one outputs a binary mask instead of a bounding box per instance). This is oftentimes called "binary mask classification". Semantic segmentation on the other hand was typically solved by making models output a single "segmentation map" with one label per pixel. Hence, semantic segmentation was treated as a "per-pixel classification" problem. An example of a popular semantic segmentation model which adopts this paradigm is [SegFormer](https://huggingface.co/docs/transformers/model_doc/segformer), on which we wrote an extensive blog post [here](https://huggingface.co/blog/fine-tune-segformer).
```
Comment (alaradirik) [inline mask2former.md:45]:
```suggestion
Over the last years, researchers have come up with several architectures that were typically very tailored to either instance, semantic or panoptic segmentation. Instance and panoptic segmentation were typically solved by outputting a set of binary masks + corresponding labels per object instance (very similar to object detection, except that one outputs a binary mask instead of a bounding box per instance). This is oftentimes called "binary mask classification". Semantic segmentation on the other hand was typically solved by making models output a single "segmentation map" with one label per pixel. Hence, semantic segmentation was treated as a "per-pixel classification" problem. An example of a popular semantic segmentation model which adopts this paradigm is [SegFormer](https://huggingface.co/docs/transformers/model_doc/segformer), on which we wrote an extensive blog post [here](https://huggingface.co/blog/fine-tune-segformer).
```

File: mask2former.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+- **instance segmentation** is the task of identifying different "instances", like individual people, in an image. Instance segmentation is very similar to object detection, except that we'd like to output a set of binary segmentation masks, rather than bounding boxes, with corresponding class labels. Instances are oftentimes also called "objects" or "things". Note that individual instances may overlap.
+- **semantic segmentation** is the task of identifying different "semantic categories", like "person" or "sky" in an image. Contrary to instance segmentation, no distinction is made between individual instances of a given semantic category; one just likes to come up with a mask for the "person" category, rather than for the individual people for example. Semantic categories which don't have individual instances, like "sky" or "grass", are oftentimes referred to as "stuff", to make the distinction with "things" (great names, huh?). Note that no overlap between semantic categories is possible, as each pixel belongs to one category.
```
Comment (alaradirik) [inline mask2former.md:42]:
```suggestion
- **semantic segmentation** is the task of identifying different "semantic categories", like "person" or "sky" of each pixel in an image. Contrary to instance segmentation, no distinction is made between individual instances of a given semantic category; one just likes to come up with a mask for the "person" category, rather than for the individual people for example. Semantic categories which don't have individual instances, like "sky" or "grass", are oftentimes referred to as "stuff", to make the distinction with "things" (great names, huh?). Note that no overlap between semantic categories is possible, as each pixel belongs to one category.
```

File: mask2former.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+Later on, Mask2Former extended this to instance segmentation by further improving the neural network architecture. Hence, we've evolved from separate architectures to what researchers now refer to as "universal image segmentation" architectures, capable of solving any image segmentation task. Interestingly, these universal models all adopt the "mask classification" paradigm, discarding the "per-pixel classification" paradigm entirely.
+
+Note that Mask2Former still needs to be trained on each task separately to obtain state-of-the-art results. This has already been improved by [OneFormer](https://arxiv.org/abs/2211.06220), which obtains state-of-the-art performance on all 3 tasks by only training on a panoptic version of the dataset, by adding a text encoder to condition the model on either "instance", "semantic" or "panoptic". This model will also become available in [`🤗 transformers`](https://huggingface.co/transformers), but comes with greater latency.
```
Comment (alaradirik) [inline mask2former.md:53]:
```suggestion
Note that Mask2Former still needs to be trained on each task separately to obtain state-of-the-art results. This has already been improved by [OneFormer](https://arxiv.org/abs/2211.06220), which obtains state-of-the-art performance on all 3 tasks by only training on a panoptic version of the dataset, by adding a text encoder to condition the model on either "instance", "semantic" or "panoptic" tasks. This model will also become available in [`🤗 transformers`](https://huggingface.co/transformers), but comes with greater latency.
```

File: mask2former.md [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+## Inference with Mask2Former in Transformers
+
+Usage of Mask2Former is pretty straightforward, and exactly the same as its predecessor MaskFormer. One prepares an image for the model using the image processor, and then forwards the `pixel_values` through the model. The model outputs a set of binary masks and corresponding class logits. Next, these can be postprocessed using the image processor to get the final instance, semantic or panoptic segmentation predictions.
```
Comment (alaradirik) [inline mask2former.md:57]: The binary mask term might confuse people as they might expect masks of 1s and 0s, just rephrasing this a bit:
```suggestion
Usage of Mask2Former is pretty straightforward, and exactly the same as its predecessor MaskFormer. One prepares an image for the model using the image processor, and then forwards the `pixel_values` through the model. The model outputs a set of binary mask and corresponding class logits. Note that the number of output binary masks is set to either 100 or 200 and each binary mask is essentially an object instance proposal which can be accepted or discarded based on the mask logits. Furthermore, the corresponding class logits are used to assign category labels to each binary mask. The raw output of Mask2Former can be easily postprocessed using the image processor to get the final instance, semantic or panoptic segmentation predictions.
```

File: mask2former.md [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+inputs = image_processor(image, return_tensors="pt")
+
+with torch.no_grad():
+    outputs = model(**inputs)
+
+predicted_panoptic_segmentation = processor.post_process_panoptic_segmentation(outputs, target_sizes=[image.size[::-1]])[0]
+```
+
+In panoptic segmentation, the final `predicted_panoptic_segmentation` contains 2 things: a segmentation map, as well as a corresponding `segments_info`, which contains more information about the individual segments of the map (such as their class ID). Here one can see that in panoptic segmentation, no overlap between segments is possible due to the fact that ground truths are stored in a single segmentation map.
```
Comment (alaradirik) [inline mask2former.md:79]:
```suggestion
In panoptic segmentation, the final `predicted_panoptic_segmentation` contains 2 things: a segmentation map of shape (height, width) where each value encodes the instance ID of a given pixel, as well as a corresponding `segments_info`. The  `segments_info` contains more information about the individual segments of the map (such as their class / category ID). Note that Mask2Former outputs binary mask proposals of shape (96, 96) for efficiency and the `target_sizes` argument is used to resize the final mask to the original image size. Here one can see that in panoptic segmentation, no overlap between segments is possible due to the fact that ground truths are stored in a single segmentation map.
```

**Review by alaradirik — APPROVED** [review]
Left a few comments to cover all bases / details but LGTM! Thanks for working on this.

File: mask2former.md [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+## Inference with Mask2Former in Transformers
+
+Usage of Mask2Former is pretty straightforward, and exactly the same as its predecessor MaskFormer. One prepares an image for the model using the image processor, and then forwards the `pixel_values` through the model. The model outputs a set of binary masks and corresponding class logits. Next, these can be postprocessed using the image processor to get the final instance, semantic or panoptic segmentation predictions.
+
+Below, we load a Mask2Former checkpoint from the hub trained on the COCO panoptic dataset (note that the authors released no less than [30 checkpoints](https://huggingface.co/models?other=mask2former)). Hence we use the `post_process_panoptic_segmentation` method to get the final predicted panoptic segmentation.
+
+```
+from transformers import AutoImageProcessor, Mask2FormerForUniversalImageSegmentation
```
Comment (shivalikasingh95) [inline mask2former.md:62]:
```suggestion
from transformers import AutoImageProcessor, Mask2FormerForUniversalSegmentation
```

File: mask2former.md [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+from PIL import Image
+
+processor = AutoImageProcessor.from_pretrained("facebook/mask2former-swin-base-coco-panoptic")
+model = Mask2FormerForUniversalImageSegmentation.from_pretrained("facebook/mask2former-swin-base-coco-panoptic")
```
Comment (shivalikasingh95) [inline mask2former.md:66]:
```suggestion
model = Mask2FormerForUniversalSegmentation.from_pretrained("facebook/mask2former-swin-base-coco-panoptic")
```

File: mask2former.md [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+predicted_panoptic_segmentation = processor.post_process_panoptic_segmentation(outputs, target_sizes=[image.size[::-1]])[0]
+```
+
+In panoptic segmentation, the final `predicted_panoptic_segmentation` contains 2 things: a segmentation map, as well as a corresponding `segments_info`, which contains more information about the individual segments of the map (such as their class ID). Here one can see that in panoptic segmentation, no overlap between segments is possible due to the fact that ground truths are stored in a single segmentation map.
+
+For visualization, we refer the reader to the [demo notebook](https://github.com/NielsRogge/Transformers-Tutorials/blob/master/Mask2Former/Inference_with_Mask2Former.ipynb), which also illustrates semantic segmentation.
+
+## Fine-tuning Mask2Former in Transformers
+
+For fine-tuning Mask2Former on a custom dataset for either instance, semantic and panoptic segmentation, we refer the reader to the [demo notebooks](https://github.com/NielsRogge/Transformers-Tutorials/tree/master/Mask2Former). Note that fine-tuning Mask2Former has the exact same API as fine-tuning MaskFormer. Hence, upgrading from MaskFormer to Mask2Former is easy!
```
[the remark this answers is not in the document]
Comment (shivalikasingh95) [reply mask2former.md:85]: I think it might be more reader friendly if we can expand a bit on how easy the switch would be rather than just saying it's easy to do.

File: mask2former.md [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+with torch.no_grad():
+    outputs = model(**inputs)
+
+predicted_panoptic_segmentation = processor.post_process_panoptic_segmentation(outputs, target_sizes=[image.size[::-1]])[0]
+```
+
+In panoptic segmentation, the final `predicted_panoptic_segmentation` contains 2 things: a segmentation map, as well as a corresponding `segments_info`, which contains more information about the individual segments of the map (such as their class ID). Here one can see that in panoptic segmentation, no overlap between segments is possible due to the fact that ground truths are stored in a single segmentation map.
+
+For visualization, we refer the reader to the [demo notebook](https://github.com/NielsRogge/Transformers-Tutorials/blob/master/Mask2Former/Inference_with_Mask2Former.ipynb), which also illustrates semantic segmentation.
```
[the remark this answers is not in the document]
Comment (shivalikasingh95) [reply mask2former.md:81]: I'm working on a space for Mask2Former so linking it here. The space is private at the moment, I'll make it public shortly before this PR gets merged.

**Comment by shivalikasingh95** [comment]
This is great, thanks for working on this Niels :) Left few small suggestions but LGTM overall!

File: mask2former.md [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+    outputs = model(**inputs)
+
+predicted_panoptic_segmentation = processor.post_process_panoptic_segmentation(outputs, target_sizes=[image.size[::-1]])[0]
+```
+
+In panoptic segmentation, the final `predicted_panoptic_segmentation` contains 2 things: a segmentation map of shape (height, width) where each value encodes the instance ID of a given pixel, as well as a corresponding `segments_info`. The  `segments_info` contains more information about the individual segments of the map (such as their class / category ID). Note that Mask2Former outputs binary mask proposals of shape (96, 96) for efficiency and the `target_sizes` argument is used to resize the final mask to the original image size. Here one can see that in panoptic segmentation, no overlap between segments is possible due to the fact that ground truths are stored in a single segmentation map.
+
+If you liked this topic and want to learn more, we recommend the following resources:
+
+- Our [demo notebook](https://github.com/NielsRogge/Transformers-Tutorials/blob/master/Mask2Former/Inference_with_Mask2Former.ipynb) notebook, that gives a broader overview on inference with Mask2Former and visualizing its results.
```
Comment (shivalikasingh95) [inline mask2former.md:83]:
```suggestion
- Our [demo notebook](https://github.com/NielsRogge/Transformers-Tutorials/blob/master/Mask2Former/Inference_with_Mask2Former.ipynb), that gives a broader overview on inference with Mask2Former and visualizing its results.
```

File: mask2former.md [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+
+Later on, Mask2Former extended this to instance segmentation by further improving the neural network architecture. Hence, we've evolved from separate architectures to what researchers now refer to as "universal image segmentation" architectures, capable of solving any image segmentation task. Interestingly, these universal models all adopt the "mask classification" paradigm, discarding the "per-pixel classification" paradigm entirely.
+
+Note that Mask2Former still needs to be trained on each task separately to obtain state-of-the-art results. This has already been improved by [OneFormer](https://arxiv.org/abs/2211.06220), which obtains state-of-the-art performance on all 3 tasks by only training on a panoptic version of the dataset, by adding a text encoder to condition the model on either "instance", "semantic" or "panoptic". This model will also become available in [`🤗 transformers`](https://huggingface.co/transformers), but comes with greater latency.
+
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply mask2former.md:76]: I added a figure of Mask2Former to make the blog post more vibrant 👍 thanks for suggesting

[the remark this answers is not in the document]
Comment (shivalikasingh95) [reply mask2former.md:76]: That's great, blog looks good!
